Adds UnsafeCollectionOperations for unsafe access to RepeatedField<T>

PascalSenn opened the pull request:
This is a proposal to add `UnsafeCollectionOperations `  for fast access on  `RepeatedField<T>`
  https://github.com/protocolbuffers/protobuf/issues/16745

jskeet commented:
@JamesNK We're aware that if the repeated field is modified while the span is "active", it may or may not reflect the changes (e.g. add/remove operations will not be seen in terms of the length, and if the array is reallocated any further updates won't be visible at all) - but do you think that's reasonable behavior, so long as we document that appropriately? (I'd just say that the state of the span is undefined after any modifications.) Any other thoughts about this PR?

JamesNK commented:
Modification is the problem. Sometimes the span reflects changes, sometimes it doesn't.

I like the idea, but I don't think it should be a property on `RepeatedField<T>`. It makes the dangerous behavior too easy to access. I suggest doing what the BCL does with `List<T>` + span and make it a static method on an "advanced" type.

https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.asspan?view=net-8.0

PascalSenn commented:
@JamesNK 

I changed the implementation to include your suggestions. There is already a similar class `UnsafeByteOperations`, so I aligned the naming with it -> `UnsafeCollectionOperations`. The implementation of the class is very close to the [CollectionsMarshal.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs) from `System.Runtime.InteropServices`. I implemented `SetCount` and `AsSpan`. I also ported the tests over from [CollectionsMarshalTests.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/CollectionsMarshalTests.cs).

jskeet commented on csharp/src/Google.Protobuf.Test/Collections/UnsafeCollectionOperationsTest.cs:
```diff
@@ -0,0 +1,210 @@
+using System;
```
Please add a copyright header (as per the other files) here.

jskeet commented on csharp/src/Google.Protobuf/Collections/RepeatedField.cs:
```diff
@@ -611,6 +600,17 @@ public override string ToString()
             }
         }
 
+        // May increase the size of the internal array, but will never shrink it.
```
Any reason this has been moved rather than just changed from private to internal?

jskeet replied:
(With my suggested approach of a new internal SetCount method, this could stay unchanged.)

PascalSenn replied:
I moved it down as it appeared to be in the wrong place for an internal method. However, after reviewing the code again, I'm not sure if there even is any specific "ordering by visibility" of members in this class.

jskeet commented on csharp/src/Google.Protobuf.Test/Collections/UnsafeCollectionOperationsTest.cs:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    [Test]
+    public void AsSpanOnConcurrentFieldAcccessThrows()
```
Acccess => Access

jskeet commented on csharp/src/Google.Protobuf/Collections/RepeatedField.cs:
```diff
@@ -34,8 +34,8 @@ public sealed class RepeatedField<T> : IList<T>, IList, IDeepCloneable<RepeatedF
 
-        private T[] array = EmptyArray;
-        private int count = 0;
+        internal T[] array = EmptyArray;
```
Please don't make fields internal. I was already nervous about having the SetCount method, but I *definitely* don't want internal access to fields.

jskeet replied:
Instead, I'd recommend adding internal SetCount and AsSpan methods, and calling those from the public methods in UnsafeCollectionOperations.

PascalSenn replied:
Yes, I understand. I follow the same patterns as in

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L25-L27

It appears to be a performance optimization in the `List<T>` class to expose these fields. 

I can revert this change.

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,145 @@
+                var size = field.count;
+                var items = field.array;
+
+                Debug.Assert(items is not null, "Implementation depends on RepeatedField<T> always having an array.");
```
Please use a normal if/throw rather than Debug.Assert. (We don't use Debug.Assert in any of the production code in protobuf.)

However, I don't think we actually need to check this anyway. I don't think the array can ever be null, and if it ever *were* null it would cause other issues.

PascalSenn replied:
resolved

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,145 @@
+        /// A <see cref="Span{T}"/> that wraps the backing array of the
+        /// <see cref="RepeatedField{T}"/>.
+        /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
I don't think we need aggressive inlining for this.

PascalSenn replied:
I believe a performance improvement from `CollectionsMarshal.cs`.
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs#L21

I removed it

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,145 @@
+        /// <remarks>
+        /// When increasing the count, uninitialized data is being exposed.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
I don't think we need aggressive inlining for this.

jskeet reviewed (commented):
I haven't reviewed the tests in detail as I'm suggesting quite significant implementation changes.

PascalSenn commented on csharp/src/Google.Protobuf.Test/Collections/UnsafeCollectionOperationsTest.cs:
```diff
@@ -0,0 +1,210 @@
+public class UnsafeCollectionOperationsTest
+{
+    [Test]
+    public unsafe void NullFieldAsSpanValueType()
```
I don't mind either change. The reasoning behind this is to follow the same patterns as the `CollectionsMarshal`.

PascalSenn replied:
I think it's more consistent to throw in case of null, as `SetCount` does the same. 
I changed it to this behaviour.

PascalSenn commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,145 @@
+                    throw new InvalidOperationException("Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.");
+                }
+
+                Debug.Assert(typeof(T[]) == field.array.GetType(), "Implementation depends on RepeatedField<T> always using a T[] and not U[] where U : T.");
```
I will remove the check as `RepeatedField<T>` is sealed anyway

PascalSenn commented:
@jskeet 
Most of this implementation closely aligned with `CollectionsMarshal.cs` from the runtime. The goal was to keep the implementation and behavior relatively consistent.

I also believe a few things were performance optimizations in the `CollectionsMarshal`. For example, the aggressive inlining and the direct field access.

I incorporated your feedback into this pull request to better align with the behavior of Protobuf. I left a couple of comments open to provide some context. Feel free to close them as needed after review, as everything should be resolved.

PascalSenn commented:
@jskeet One last remark. The tests are ported from `CollectionsMarshalTests.cs`   to ensure consistent behavior:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/CollectionsMarshalTests.cs

jskeet commented:
I'll have a look tomorrow. I would say that I'm not as interested in alignment with CollectionsMarshal as with consistency with the rest of *this* codebase, and clarity of code and usage.

JamesNK commented on csharp/src/Google.Protobuf.Test/Google.Protobuf.Test.csproj:
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>net462;net6.0</TargetFrameworks>
```
How come net462 is removed? I think it's valuable to run tests there. If some features aren't supported on .NET Framework then they can be selectively `#ifdef`ed out.

PascalSenn replied:
Oops. I removed it to be able to run the test on my unix machine

JamesNK commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,118 @@
+
+namespace Google.Protobuf
+{
+    /// <summary>
```
I think the info here should be merged with the comment on `AsSpan`.

The overall comment for the type could be similar to CollectionMarshal: https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal?view=net-8.0

> An unsafe class that provides a set of methods to access the underlying data representations of collections.

The reason to simplify it is because in the further there could be methods that are specific to `MapField`, and then the current class comment wouldn't make sense.

jskeet replied:
+1

JamesNK commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,118 @@
+    public static class UnsafeCollectionOperations
+    {
+        /// <summary>
+        /// Returns a <see cref="Span{T}"/> that wraps the backing array of the specified
```
The comment should mention:

Values in the `Span<T>` shouldn't be set to `null`. Items should be removed from a RepeatedField with Remove or RemoveAt.

jskeet replied:
Possibly change "the backing array" to "the current backing array" to emphasize that it can change?

JamesNK reviewed (commented):
I think its good! Some minor comment and test improvements in feedback

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,118 @@
+using System.Security;
+using Google.Protobuf.Collections;
+
+namespace Google.Protobuf
```
Let's use a file-scoped namespace for simplicity.

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,118 @@
+        /// </exception>
+        public static void SetCount<T>(RepeatedField<T> field, int count)
+        {
+            if (field is null)
```
Use `ProtoPreconditions.CheckNotNull(field, nameof(field))` to simplify this and the check above.

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,118 @@
+                throw new ArgumentNullException(nameof(field));
+            }
+
+            if (count < 0)
```
I'd expect this test to be in RepeatedField.SetCount instead. Basically this method can just be a proxy.

jskeet reviewed (commented):
Down to just nits now - it feels like we're very close, thanks very much for this!

jskeet commented:
cc @amanda-tarafa for any additional thoughts

amanda-tarafa reviewed (commented):
A couple of comments, but looks good otherwise.

jskeet commented:
Thanks - I'm afraid at this point I won't get to this until Monday, but hopefully I'll be able to look then.

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,88 @@
+    /// remove items instead.
+    /// </para>
+    /// <para>
+    /// The returned <see cref="Span{T}"/> is only valid until the <see cref="RepeatedField{T}"/> is
```
This isn't quite right - it's fine for the field to be modified so long as the size doesn't change. (So modifying existing elements preserves the link.)

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,88 @@
+    /// The type of elements in the <see cref="RepeatedField{T}"/>.
+    /// </typeparam>
+    /// <param name="field">The field to set the count of.</param>
+    /// <param name="count">The value to set the field's count to.</param>
```
Specify "Must be non-negative."

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,88 @@
+    /// <typeparam name="T">
+    /// The type of elements in the <see cref="RepeatedField{T}"/>.
+    /// </typeparam>
+    /// <param name="field">The field to set the count of.</param>
```
Specify "Must not be null."

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,88 @@
+    /// The type of elements in the <see cref="RepeatedField{T}"/>.
+    /// </typeparam>
+    /// <param name="field">
+    /// The <see cref="RepeatedField{T}"/> for which to wrap the current backing array.
```
Specify "Must not be null."

jskeet replied:
(We haven't historically been terribly good at this in the proto library, but I'd like to try to make sure that at least new code is clearer...)

jskeet commented on csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs:
```diff
@@ -0,0 +1,88 @@
+    /// <param name="field">The field to set the count of.</param>
+    /// <param name="count">The value to set the field's count to.</param>
+    /// <exception cref="ArgumentNullException">
+    /// Thrown if <paramref name="field"/> is <see langword="null"/>.
```
Specify ArgumentOutOfRangeException as well? (If we're doing one, we might as well do both.)

PascalSenn replied:
I added the comment, though the precondition check was moved to `SetCount` on `RepeatedField`.

jskeet reviewed (commented):
Just down to comment nits from my side.

PascalSenn commented:
@jskeet  Comments are resolved 🚀

jskeet commented:
@JamesNK: Could you take one final pass at this? It looks good to me, but due to the unsafe aspects of it, I'd appreciate a double-check. After that, I can get it reviewed by a member of the protobuf team and merged.

jskeet commented:
@mkruskal-google: We've reviewed this from a .NET perspective. Let me know if you'd like any more background. (But basically this allows very efficient access to a repeated field, with caveats around anything that might resize it.)

mkruskal-google commented:
This lgtm, @PascalSenn we need you to sign the CLA though

PascalSenn commented:
@mkruskal-google CLA Is signed.
The failed test is somehow related to the `[AllowPartiallyTrustedCallers]`. It's a attribute from .NET Framework. I work on a unix machine and cannot run the windows tests. @jskeet / @JamesNK  Do i just have to add `[AllowPartiallyTrustedCallers]` to the `UnsafeCollectionOperations` aswell?

jskeet commented:
@PascalSenn: It's always been an area I really find hard to understand, to be honest. @JamesNK could you advise, please?

JamesNK commented:
You need to put `[SecuritySafeCritical]` attribute on methods that return span. See https://github.com/protocolbuffers/protobuf/issues/7700 as an example

JamesNK commented:
I copied your branch locally and ran net462 tests. Here are the changes you need to make:

https://github.com/JamesNK/protobuf/commit/55dda6293ff82291288d630158499b9fec017d00

PascalSenn commented:
Thanks @jskeet  & @JamesNK, tests should pass now.

jskeet commented:
Ping @mkruskal-google for approval.

verdie-g commented:
Hi, I would be very interested by that feature too. Is there any chance to go forward with that PR?

JasonLunn commented:
@JamesNK - you're being flagged for not having signed the CLA but having contributed to the last commit. Can you resolve that?

JamesNK commented:
I'm pretty sure I already have. When I click on the CLA details link I see this:

![image](https://github.com/user-attachments/assets/48046ade-041e-477c-89a6-6a603461b928)

JasonLunn commented:
> I'm pretty sure I already have. When I click on the CLA details link I see this:
> 
> ![image](https://private-user-images.githubusercontent.com/303201/375924261-48046ade-041e-477c-89a6-6a603461b928.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg3MzU4MjQsIm5iZiI6MTcyODczNTUyNCwicGF0aCI6Ii8zMDMyMDEvMzc1OTI0MjYxLTQ4MDQ2YWRlLTA0MWUtNDc3Yy04OWE2LTZhNjAzNDYxYjkyOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxMlQxMjE4NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hYjMwOTM4YWVkYzY5MWI4NDgyOTU3MDIxOWU4MzIxZGY2NTU1NWM3MGI5NzIzMGY2NGEyMGVjZmNkMGRiZjQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.A8d-5xom6HxpU2XukVSwzSKq7E_Y_zTBSZenRjnZ0T8)

Can you check that the contact info matches `j***s​@newtonking.com` - that's the email associated with commit https://github.com/protocolbuffers/protobuf/commit/46ed223b2c3aceece6d2128c258bb94bedf8fd26 which is the one in question.

JamesNK commented:
Ok, the signed in google account was different. I signed it with a different email address.

JasonLunn commented:
@PascalSenn - could you push a rebase? All the test failures are outside of the C# directory, I suspect their all due to drift with test targets given this PR's longevity.

prezaei commented:
@JamesNK , while you are at it, can you please add an overload for AddRange that accepts ReadOnlyMemory<T> or similar?
